### ISSUE TITLE
RFC + code: better default 'type' for responses  

### ISSUE THREAD

**Issue by mikemaccana** [arc-repos/architect#244]
Talked about this a little in https://github.com/arc-repos/architect/issues/220 but wanted to raise it as a separate idea because I think it should be in architect itself:

## Some observations

- Typing `text/html; charset=UTF-8` (and the various other content types) is repetitive, makes code less interesting to read, and more prone to failure.

- It's very easy for us to distinguish from HTML, text and JSON and set the content type accordingly if it's omitted

## RFC

Routes should be able to return arc responses just like they do now.

But also:

```
return {
  status: 403,
  body: `
    <p>hello world<p>
  `
}

```

should return a 403 Forbidden response with `type: text/html; charset=UTF-8` - because the body is a String and the first non-whitespace character is a `<`

But also:

```
return {
  status: 200,
  body: 'hello world'
}
```

should return a 200 OK response with `type: text/plain; charset=UTF-8` - because the body is a String and the first non-whitespace character is not a '<' 

```
return {
  status: 201,
  body: {
    dinosaurID: 26
  }
}
```

should return a 201 created response with `type: application/json` and the body being the JSON encoded object - because the body is an Object.

## In other words

Routes should be able to return strings, and the content type assigned be either JSON for objects, and for Strings, HTML or test based on the first non-whitespace character

**Comment by brianleroux** [comment]
This is pretty dope, there's two ways, either in JavaScript or with Velocity Templates in API Gateway. The former is a lot easier. The latter would be difficult, but possible! Not sure how this would impact other runtime languages either. We could prototype in `arc.middleware` tho I like the idea of leaving middleware agnostic of req/res schemas. Maybe we can make it smart enough to 'know' if its middleware for HTTP, SNS or SQS…

**Comment by mikemaccana** [comment]
Thanks! Looking at doing this in JavaScript. I could throw it in arc.middleware, but it would be better to be seamless, eg. just make `type` optional in the returned `response`. I'm having trouble finding the right point in `@architect/architect` to inject this though - I want to put it somewhere that's agnostic of the sandbox / aws / future cloud providers. Where would be the best function/method to add this?

**Comment by ryanblock** [comment]
This is going to go in one of two places (correct me if I'm wrong, @brianleroux!):
- API Gateway / dep-free: gonna require some magic in the [response VTL template](https://github.com/arc-repos/architect/blob/231/src/create/aws/_create-route/_create-resource/_response.vtl), [more here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html)
- JS via dependency: `@architect/functions` -[likely here](https://github.com/arc-repos/arc-functions/blob/master/src/http/index.js#L62)

**Comment by mikemaccana** [comment]
Thanks Ryan. The VTL is specific to AWS and doesn't affect the sandbox right? I was hoping for some shared code between AWS and the sandbox - maybe in/before the validation of responses. I've been trying to break AWS with bad responses to see where it errors but AWS just seems tolerant of bad (missing keys, wrong types etc) responses.

**Comment by ryanblock** [comment]
Yeah, forgot to add the `sandbox` impl point, which [should be here](https://github.com/arc-repos/architect/blob/107a7608b10cd209bb90c10a49d839841a3d20c3/src/sandbox/http/register-route.js#L40).

And yes, we aren't attempting to mock AWS-flavored VTL in `sandbox` (for better or worse, I'd argue vastly for the better), so they wouldn't share code.

**Comment by mikemaccana** [comment]
Ah so AWS and sandbox are implemented seperately, and there's no shared
validation code for responses.

Hrm. Learning VTL sounds like a time suck and adding it to functions
defeats the purpose of making responses simpler (since you now have to run
a function rather than just return an object).

Maybe (since people might mess up responses on AWS) spme shared (basic)
response validation code should exist?

**Comment by mikemaccana** [comment]
Example ways to mess up a response: return a non existant status code,
return an array for an HTTP header value, set location for a redirect to be
undefined, etc.

I don't know enough about VTL and what it does. I just imagine there would
need to be some validation in JS to ensure arc user's  errors are caught
and users get good error messages back.

**Comment by mikemaccana** [comment]
Jumped on a laptop (it's nearly midnight here) to check out the `vtl` in more detail. 

OK responses can be manipulated in VTL on AWS, and for in JS around the implementation point @ryanblock  mentioned for sandbox. Leave this with me and I'll check it out further.

**Comment by ryanblock** [comment]
Hey, I was in our VTL stuff working working on something else and was able to get content type detection working. Are we still interested? If so, what are the rules?

As a reminder: not specifying content type currently defaults to `application/json` (and does not specify `charset=utf-8`).

Some ideas in pseudocode:
- `if (!type && startsWith('<')) type = 'text/html; charset=utf-8;'`
- `if (!type) type = 'application/json; charset=utf-8;'`

Not sure if it would be reliable or desirable to do any other content types.

`sandbox` mocking of the above should be trivial though.

**Comment by ryanblock** [comment]
Closing due to lack of interest.

**Comment by mikemaccana** [comment]
Sorry. I had this complete in JavaScript but I have no idea about how to get started with  VTL and it seems like a quagmire.

**Comment by ryanblock** [comment]
np, I think I can implement it in VTL, go ahead reopen if / when you want to post up your code and take another look.

**Comment by mikemaccana** [comment]
Thanks Ryan. Will do.
